kong/plugins/pre-function/_handler.lua

```diff
@@ -3,9 +3,73 @@
 
 local config_cache do
 
+  local env = {}
 
   local no_op = function() end
 
+  local function make_read_only_table(t)
+    setmetatable(t, {
+      __newindex = function()
+        error("This table is read-only.")
+      end
+    })
+  end
+
+  -- safe Lua modules and functions
+  ([[
+  _VERSION assert error    ipairs   next pairs
+  pcall    select tonumber tostring type unpack xpcall
+  coroutine.create coroutine.resume coroutine.running coroutine.status
+  coroutine.wrap   coroutine.yield
+  math.abs   math.acos math.asin  math.atan math.atan2 math.ceil
+  math.cos   math.cosh math.deg   math.exp  math.fmod  math.floor
+  math.frexp math.huge math.ldexp math.log  math.log10 math.max
+  math.min   math.modf math.pi    math.pow  math.rad   math.random
+  math.sin   math.sinh math.sqrt  math.tan  math.tanh
+  os.clock os.difftime os.time
+  string.byte string.char  string.find  string.format string.gmatch
+  string.gsub string.len   string.lower string.match  string.reverse
+  string.sub  string.upper
+  table.insert table.maxn table.remove table.sort
+  ]]):gsub('%S+', function(id)
+    local module, method = id:match('([^%.]+)%.([^%.]+)')
+    if module then
+      env[module]         = env[module] or {}
+      env[module][method] = _G[module][method]
+    else
+      env[id] = _G[id]
+    end
+  end)
+
+  make_read_only_table(env.coroutine)
+  make_read_only_table(env.math)
+  make_read_only_table(env.os)
+  make_read_only_table(env.string)
+  make_read_only_table(env.table)
+
+
+  local function make_function(func_str)
+    local f = assert(loadstring(func_str))
```

Comment:
```suggestion
    local f = assert(loadstring(func_str, nil, "t"))
```
to avoid loading bytecode